src/librustc/ty/layout.rs

```diff
@@ -573,12 +572,8 @@ impl<'a, 'gcx, 'tcx> Struct {
         // Neither do  1-member and 2-member structs.
         // In addition, code in trans assume that 2-element structs can become pairs.
         // It's easier to just short-circuit here.
-        let mut can_optimize = (fields.len() > 2 || StructKind::EnumVariant == kind)
-            && ! (repr.c || repr.packed);
-
-        // Disable field reordering until we can decide what to do.
-        // The odd pattern here avoids a warning about the value never being read.
-        if can_optimize { can_optimize = false; }
+        let can_optimize = (fields.len() > 2 || StructKind::EnumVariant == kind)
+            && ! (repr.c || repr.packed || repr.linear || repr.simd);
```

Comment:
We didn't optimize for `repr(simd)` before.  There might be a case for removing it from the condition.  Simd vectors (used to?) go through `Layout::Vector`.  I'm not as familiar with the details as I used to be: I had to leave this for a month in the middle.  But I think leaving it for clarity is fine.

`repr(linear)` is not a thing externally.  It exists because we need to be able to remember which structs reordering was disabled for when using optimization fuel.  It is so named because there has been talk of making an RFC that adds it, and at this point it would just be exposing it via the parser.